local-storage/src/provider.rs

```diff
@@ -75,13 +77,64 @@ impl SqliteStorageProvider {
 
         Ok(())
     }
+
+    fn get_blocks_recursive_query(
+        &self,
+        cid: &str,
+        offset: Option<u32>,
+        window_size: Option<u32>,
+    ) -> Result<Vec<StoredBlock>> {
+        let mut base_query = "
+        WITH RECURSIVE cids(x,y,z) AS (
+            SELECT cid,data,filename FROM blocks WHERE cid = (?1)
+            UNION
+            SELECT cid,data,filename FROM blocks b 
+                INNER JOIN links l ON b.cid==l.block_cid 
+                INNER JOIN cids ON (root_cid=x)
+        )
+        SELECT x,y,z FROM cids
+        "
+        .to_string();
+        let mut params = vec![cid.to_string()];
+
+        if let Some(offset) = offset {
+            if let Some(window_size) = window_size {
+                base_query.push_str(" LIMIT (?2) OFFSET (?3);");
+                params.push(format!("{window_size}"));
+                params.push(format!("{offset}"));
+            }
+        }
+        let params = params_from_iter(params.into_iter());
+        let blocks = self
+            .conn
+            .prepare(&base_query)?
+            .query_map(params, |row| {
+                let cid_str: String = row.get(0)?;
+                let data: Vec<u8> = row.get(1)?;
+                let filename: Option<String> = row.get(2).ok();
+                let links = match self.get_links_by_cid(&cid_str) {
+                    Ok(links) => links,
+                    Err(_) => vec![],
+                };
+                Ok(StoredBlock {
+                    cid: cid_str,
+                    data,
+                    links,
+                    filename,
+                })
+            })?
+            .filter_map(|b| b.ok())
+            .collect();
+
+        Ok(blocks)
+    }
 }
 
 impl StorageProvider for SqliteStorageProvider {
     fn import_block(&self, block: &StoredBlock) -> Result<()> {
         self.conn.execute(
-            "INSERT OR IGNORE INTO blocks (cid, data) VALUES (?1, ?2)",
-            (&block.cid, &block.data),
+            "INSERT OR IGNORE INTO blocks (cid, data, filename) VALUES (?1, ?2, ?3)",
```

Comment:
The "or ignore" makes me immediately question what the PK and/or unique constraints are. And looking at that code leads me to a side question: The purpose of an autonumber-style "id"? 

But back to the main question: it looks like you did not modify the unique index on CID. So if you add a second file with the same content it gets silently dropped completely - no record of that filename existing. Is this really desired?

Perhaps it's too academic a concern, but in a pure mathematical sense a filename probably should't be a column in blocks, right? It would be a separate table that associates... I guess many-to-many?